docs/vivisect/devguides/vstruct.rst

```diff
@@ -0,0 +1,104 @@
+.. _vstruct:
+
+Structure Parsing
+#################
+
+Parsing a series of contiguous bytes into a single structure is a common task in binary analysis. To facilitate easier parsing and present a unified API throughout the vivisect/vdb codebase, vivisect also contains the vstruct module, which is an API for automatically parsing out bytes into python objects.
+
+Typical usage of a vstruct structure definition is instantiating an instance of a vstruct object, pointing it at a set of bytes, and voila, you should have a fully populated structure. So something like this works::
+
+    >>> import vstruct.defs.pe as vs_pe
+    >>> byts = b'.data\x00\x00\x00\x01\xc0\x00\x00\xcd\xab\x00\x00\xaa\xaa\x00\x00\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xd3<\x0f\x00\x00\x00\x00\x00'
+    >>> header = vs_pe.IMAGE_SECTION_HEADER()
+    >>> header.vsParse(byts)
+
+Now, in the interest of full honesty, those bytes weren't actually handcrafted. In addition to structure parsing, vstruct classes can also just emit the bytes that comprise them. So in reality, I generated those bytes like so::
+    
+    >>> header = vs_pe.IMAGE_SECTION_HEADER()
+    >>> header.vsSetField('VirtualSize', 0xC001)
```

Comment:
you use `vsSetField()` for everything?  or are you just trying to show that you can?

i would use"
`header.VirtualSize = 0xc001` :)  it's the beauty of VStruct.

Reply:
For both this comment and the one below, yes (to your first comment). This was meant more as an illustrative example to get people started. While I know you can just do the assignment, I wanted to write down something that looked and felt like a real API. I didn't really want to write down every single way you could do something, since that gets eye-glazing-over wordy since I'd have to explain the setattr override we've got going on.

But I'll add a short blurb on it, though I do want to avoid the "Here's literally everything you could possibly do with everything" because I don't want to write full refs docs quite yet.

Reply:
i completely understand.  we don't want to tell everything about everything.

but this is the core part of VStruct.  not an afterthought.  if anything, what you have here is "the afterthought"... the magic underpinning the way Visi wanted to interact with VStructs.  VStruct was created to allow the easiest "object-like" interactions and the ability to emit and parse easily.  you may want to throw in something about the setattr override, but literally it should probably be:

```In order to make VStructs super easy to interact with just like any ordinary object, we've used a little Python magic (overriding setattr/getattr if you're curious) to allow you to work like this:
   structhing.barf = 20
   print("you don't want to barf: %d" % structhing.barf)
```

you're teaching here.  people will think your way is the "right way" when is not (at least, not my understanding of what is "right").

Reply:
i completely understand. we don't want to tell everything about everything.

but this is the core part of VStruct. not an afterthought. if anything, what you have here is "the afterthought"... the magic underpinning the way Visi wanted to interact with VStructs. VStruct was created to allow the easiest "object-like" interactions and the ability to emit and parse easily. you may want to throw in something about the setattr override, but literally it should probably be:

```
   structhing.barf = 20
   print("you don't want to barf: %d" % structhing.barf)
```
you're teaching here. people will think your way is the "right way" when is not (at least, not my understanding of what is "right").

Reply:
I would hardly call this the "wrong" way to do things, since the setattr override calls the vsAddField/vsSetField functions anyway, and it is important to note the differences between the two.

Actually, this is how I first figured out vstruct, so I consider the setattr overrides to be convenience functions that sit atop the vsAdd/Set methods that form the core API. So right/wrong are subjective in this case. Both APIs work and are meant to be used, and I added a blurb about the setattr override.

Reply:
you see, you "figured out" vstruct, which means that you approached it like learning some object-y thing.

i'm not calling this way wrong... it clearly works, and is an important part of the implementation upon which the set/getattr hooks make use of.  it's just not the way it was intended to be used.  you showing that as the only example would leave me personally going "wtfo?  why do i care?  i'll just use struct.(un)pack and be done."  these "helper" functions set the tone for using vstruct, at least for me.

Reply:
I would still say that a lot of this is both subjective (you and I have noted that we approach problems very differently, which isn't wrong) and moot (since both are in the docs now).